File: python/src/problems1_100/merge_two_sorted_lists.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode


class Problem(ABC):
    @abstractmethod
    def merge_two_lists(
        self,
        l1: SinglyListNode | None,
        l2: SinglyListNode | None,
    ) -> SinglyListNode | None:
        pass


class Default(Problem):
    @override
    def merge_two_lists(
        self,
        l1: SinglyListNode | None,
        l2: SinglyListNode | None,
    ) -> SinglyListNode | None:
        if not l1:
            return l2
        if not l2:
            return l1
        root = SinglyListNode()
        current = root
        # when both items are there
        while l1 and l2:
            if l1.value < l2.value:
                current.next = SinglyListNode(l1.value)
                l1 = l1.next
            else:
                current.next = SinglyListNode(l2.value)
                l2 = l2.next
            current = current.next
        # remaining
        current.next = l1 or l2
        return root.next


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems1_100/merge-two-sorted-lists.js
import SinglyListNode from '../concepts/singly-list-node';

class Problem {
  mergeTwoLists(l1, l2) {
    throw new Error(`Not implemented (${l1}, ${l2}).`);
  }
}

class Default extends Problem {
  mergeTwoLists(l1, l2) {
    if (!l1) {
      return l2;
    }
    if (!l2) {
      return l1;
    }
    const root = new SinglyListNode();
    let current = root;

    // when both items are there
    while (l1 && l2) {
      if (l1.val < l2.val) {
        current.next = new SinglyListNode(l1.val);
        l1 = l1.next;
      } else {
        current.next = new SinglyListNode(l2.val);
        l2 = l2.next;
      }
      current = current.next;
    }

    // remaining
    current.next = l1 || l2;
    return root.next;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
